### PR TITLE
Adds a new proc to objects - is_hot()

### PR DESCRIPTION
**Pull request by spaceplaceexplorer**
Objects now have a new proc `is_hot()` and var heat to determine whether they should ignite a mob upon hitting it.

This is checked when getting hit by a projectile in `bullet_act()`, melee in `apply_hit_effect()`, and thrown object in `hitby()`.

Added a new define for `HEAT_MOBIGNITE_THRESHOLD 400`, which is the minimum amount of heat needed to call `IgniteMob()` when hit. By itself, this does not cause fires, only if the target already has fire_stacks present.

All burn damage energy and beam projectiles cause such heat ignition, so do many items such as cigs, candles, matches, flares and welders.

### CODE REVIEW

File: code/game/objects/items.dm [hunk 1]
```diff
@@ -461,6 +461,11 @@ modules/mob/living/carbon/human/life.dm if you die, you will be zoomed out.
 //Simple form ideal for basic use. That proc will return TRUE only when everything was done right, and FALSE if something went wrong, ot user was unlucky.
 //Editionaly, handle_failure proc will be called for a critical failure roll.
 /obj/item/proc/use_tool(var/mob/living/user, var/atom/target, base_time, required_quality, fail_chance, required_stat = null, instant_finish_tier = 110, forced_sound = null)
+	if (required_quality == QUALITY_WELDING)
```
Comment (NanakoAC) [inline code/game/objects/items.dm]: use your new variable here instead of qualities.

For example, a laser scalpel or an electric drill mitght also ignite

File: code/game/objects/items/weapons/tools/_tools.dm [hunk 2]
```diff
@@ -217,31 +217,29 @@
 				spawn(100)
 					H.disabilities &= ~NEARSIGHTED
 
-//Prosthetic repair
-/obj/item/weapon/tool/attack(var/mob/living/carbon/human/H, var/mob/living/user)
 
-	if(get_tool_type(user, list(QUALITY_WELDING)))
-		if(ishuman(H))
-			var/obj/item/organ/external/S = H.organs_by_name[user.targeted_organ]
+/obj/item/weapon/tool/attack(mob/living/M, mob/living/user, var/target_zone)
```
Comment (NanakoAC) [inline code/game/objects/items/weapons/tools/_tools.dm]: this section seems a bit out of scope, what's this all about?

File: code/game/objects/items/weapons/tools/weldingtools.dm [hunk 3]
```diff
@@ -31,6 +31,7 @@
 		..()
 		damtype = BURN
 		force = force_ignited
+		ignites_with_touch = 1
```
Comment (NanakoAC) [inline code/game/objects/items/weapons/tools/weldingtools.dm]: use TRUE/FALSE instead of 1/0

File: code/game/objects/objs.dm [hunk 4]
```diff
@@ -12,6 +12,7 @@
 	var/damtype = "brute"
 	var/armor_penetration = 0
 	var/corporation = null
+	var/ignites_with_touch //if 1, ignites things when it hits them, does not add fire stacks by itself
```
Comment (NanakoAC) [inline code/game/objects/objs.dm]: Give it an initial value of FALSE

File: code/modules/mob/living/carbon/human/human_defense.dm [hunk 5]
```diff
@@ -290,6 +290,9 @@ meteor_act
 		var/obj/item/organ/external/affecting = get_organ(zone)
 		var/hit_area = affecting.name
 
+		if (O.ignites_with_touch)
+			IgniteMob()
```
Comment (NanakoAC) [inline code/modules/mob/living/carbon/human/human_defense.dm]: does this proc check the fire stacks?

File: code/game/objects/objs.dm (hunk 4 above)
Comment (spaceplaceexplorer) [reply code/game/objects/objs.dm]: Why?

File: code/game/objects/items/weapons/tools/weldingtools.dm (hunk 3 above)
Comment (spaceplaceexplorer) [reply code/game/objects/items/weapons/tools/weldingtools.dm]: The variable can be potentially used to represent temperature for ignition purposes. I use 1 and 0 everywhere for clarity since there is no boolean type in byond and it is misleading.

File: code/modules/mob/living/carbon/human/human_defense.dm (hunk 5 above)
Comment (spaceplaceexplorer) [reply code/modules/mob/living/carbon/human/human_defense.dm]: IgniteMob only ignites when there are fire_stacks present.

File: code/game/objects/items/weapons/tools/_tools.dm (hunk 2 above)
Comment (spaceplaceexplorer) [reply code/game/objects/items/weapons/tools/_tools.dm]: The prosthetic repair was not resolving the attack properly and under certain circumstances just returning without feedback so I fixed the conditions and returns.

**Comment by NanakoAC** [comment]
and flares

**Comment by LudwigVonChesterfield** [comment]
Why not have a “can_ignite()” proc instead of a var you need to keep of all the time, toggle when lit/extinguished, and etc.?

**Comment by spaceplaceexplorer** [comment]
How would you imagine such a proc to work? I'm not saying it's a bad idea.

**Comment by spaceplaceexplorer** [comment]
Changed the var into a proc

File: code/game/objects/items/weapons/cigs_lighters.dm [hunk 6]
```diff
@@ -16,17 +16,23 @@ CIGARETTE PACKETS ARE IN FANCY.DM
 /obj/item/weapon/flame
 	var/lit = 0
 
+/obj/item/weapon/flame/ignites_with_touch()
+	if (lit)
```
Comment (LudwigVonChesterfield) [inline code/game/objects/items/weapons/cigs_lighters.dm]: Could write return lit instead.

File: code/game/objects/items/weapons/tools/weldingtools.dm [hunk 7]
```diff
@@ -55,3 +55,7 @@
 	switched_on_qualities = list(QUALITY_WELDING = 40, QUALITY_CAUTERIZING = 15, QUALITY_WIRE_CUTTING = 15)
 	max_fuel = 40
 	force_ignited = WEAPON_FORCE_PAINFULL*1.15 //Slightly more powerful, not much more so
+
+/obj/item/weapon/tool/weldingtool/ignites_with_touch()
+	if (damtype == BURN)
```
Comment (LudwigVonChesterfield) [inline code/game/objects/items/weapons/tools/weldingtools.dm]: Doesn’t it have a lit var?

Comment (spaceplaceexplorer) [reply code/game/objects/items/weapons/tools/weldingtools.dm]: It does not.

**Comment by spaceplaceexplorer** [comment]
All requested changes now included.

File: code/game/objects/items/weapons/cigs_lighters.dm (hunk 6 above)
Comment (spaceplaceexplorer) [reply code/game/objects/items/weapons/cigs_lighters.dm]: It now returns heat.

**Comment by LethalGhost** [comment]
Proc called `is_hot` should return TRUE/FALSE.
`get_temperature` is more suitable name

File: code/game/objects/objs.dm [hunk 8]
```diff
@@ -12,8 +12,10 @@
 	var/damtype = "brute"
 	var/armor_penetration = 0
 	var/corporation = null
+	var/heat = 0
 
-
+/obj/proc/is_hot()
+	return heat
```
Comment (LethalGhost) [inline code/game/objects/objs.dm]: That really should be `/obj` level proc/var?

Comment (spaceplaceexplorer) [reply code/game/objects/objs.dm]: All objects can collide with mobs. Therefor all objects should have such property.

File: code/game/objects/items/devices/lighting/flare.dm [hunk 9]
```diff
@@ -10,6 +10,7 @@
 	var/on_damage = 7
 	var/produce_heat = 1500
 	turn_on_sound = 'sound/effects/Custom_flare.ogg'
+	heat = 1000
```
Comment (NanakoAC) [inline code/game/objects/items/devices/lighting/flare.dm]: set to 1873, flares burn reallly hot

File: code/game/objects/items/weapons/cigs_lighters.dm [hunk 10]
```diff
@@ -15,6 +15,11 @@ CIGARETTE PACKETS ARE IN FANCY.DM
 //For anything that can light stuff on fire
 /obj/item/weapon/flame
 	var/lit = 0
+	heat = 1000
```
Comment (NanakoAC) [inline code/game/objects/items/weapons/cigs_lighters.dm]: 1670, temperature of a burning candle

File: code/game/objects/items/weapons/tools/weldingtools.dm [hunk 11]
```diff
@@ -23,6 +23,8 @@
 	create_hot_spot = TRUE
 	glow_color = COLOR_ORANGE
 
+	heat = 3800
```
Comment (NanakoAC) [inline code/game/objects/items/weapons/tools/weldingtools.dm]: 2250 for normal welder, 3773 for advanced one

File: code/modules/mob/living/living_defense.dm [hunk 12]
```diff
@@ -50,6 +50,8 @@
 
 
 /mob/living/bullet_act(var/obj/item/projectile/P, var/def_zone)
+	if (P.is_hot() >= HEAT_MOBIGNITE_THRESHOLD)
```
Comment (NanakoAC) [inline code/modules/mob/living/living_defense.dm]: do projectiles work ? IIRC you said they didn't.

Can you ignite someone by dousing them in flame and firing a laser at them?

File: code/modules/overmap/ships/computers/sensors.dm [hunk 13]
```diff
@@ -36,7 +36,7 @@
 		data["range"] = sensors.range
 		data["health"] = sensors.health
 		data["max_health"] = sensors.max_health
-		data["heat"] = sensors.heat
+		data["heat"] = sensors.current_heat
```
Comment (NanakoAC) [inline code/modules/overmap/ships/computers/sensors.dm]: what's the changes to this file all about?

File: code/modules/projectiles/projectile.dm [hunk 14]
```diff
@@ -72,6 +72,10 @@
 	var/matrix/effect_transform			// matrix to rotate and scale projectile effects - putting it here so it doesn't
 										//  have to be recreated multiple times
 
+/obj/item/projectile/is_hot()
+	if (damage_type == BURN)
+		return damage*heat
```
Comment (NanakoAC) [inline code/modules/projectiles/projectile.dm]: this will produce completely insane values, best not. Just go with 3800 as a flat value for lasers

File: code/modules/projectiles/projectile/beams.dm [hunk 15]
```diff
@@ -14,6 +14,8 @@
 	tracer_type = /obj/effect/projectile/laser/tracer
 	impact_type = /obj/effect/projectile/laser/impact
 
+	heat = 200
+
 /obj/item/projectile/beam/practice
```
Comment (NanakoAC) [inline code/modules/projectiles/projectile/beams.dm]: zero out the heat for practise beams. they're basically just harmless holograms

File: code/modules/mob/living/living_defense.dm (hunk 12 above)
Comment (spaceplaceexplorer) [reply code/modules/mob/living/living_defense.dm]: They work now and yes you can.

File: code/modules/overmap/ships/computers/sensors.dm (hunk 13 above)
Comment (spaceplaceexplorer) [reply code/modules/overmap/ships/computers/sensors.dm]: That object was already using the heat war so I renamed it to current_heat instead. It was only used locally by that object.

File: code/modules/projectiles/projectile/beams.dm (hunk 15 above)
Comment (spaceplaceexplorer) [reply code/modules/projectiles/projectile/beams.dm]: It is already 0 if they deal 0 damage.
